sponsors/models.py

```diff
@@ -819,3 +839,33 @@ def nullify(self, commit=True):
         if commit:
             self.sponsorship.save()
             self.save()
+
+
+class BenefitFeature(PolymorphicModel):
+    benefit = models.ForeignKey(SponsorshipBenefit, on_delete=models.CASCADE)
```

Comment:
Thinking now, we will likely need a set of "matching models" for relation to "SponsorBenefit" objects that mirror the base... similar to how we mirror "SponsorshipBenefit" data into instances of "SponsorBenefit".

Specifically so that values can be customized after implementation. I need to think about this, but for now I think we are on the right course for Sponsor Logo placements.

It may make sense to have a "BenefitFeatureConfiguration" that is associated with the SponsorshipBenefit... and then individual "BenefitFeature" instances that exist associated with individual SponsorBenefits.

Reply:
Hm, I see your point here! What do you think about having a [JSON field](https://docs.djangoproject.com/en/3.2/ref/contrib/postgres/fields/#jsonfield) on the sponsor benefit listing all the features that were applied when it was created. Something like this:

```python
sponsor_benefit.feature_configuration = [
  {
    "benefit_model": "sponsors.models.SponsorLogoPlacement",
    "benefit_kwargs": {"publisher": "psf", "logo_place": "sponsors"},
  }
]
```

That way we'll be able to always initialize all the features related to the sponsors considering the values that were defined at creation time. This is pretty much the same idea you suggested with the`BenefitFeatureConfiguration` model, but without having an extra model.

My concern on duplicating models is that we'll have duplicate the one-level tree of specific benefit features as well. For example, for the logo placement will have to have a `LogoPlacementeConfiguration` (FK to `SponsorshipBenefit`) and a `LogoPlacement` (FK to `SponsorBenefit`). 

This duality also brings a new questions when it comes to specific's feature behavior implementation. For example, we'll probably have to implement a method under the logo placement model to define things such as the placement type (image only or image + text) or run validations. Where we should place such behavior if we have 2 very similar classes? It's highly probable that the right place would be the `SponsorBenefit` related class, but I don't have general understanding of upcoming features to assert that this will always be true.

The downside I see when it comes to this JSON strategy is that refactoring, such as renames, on the features models will come with a high risky of breaking how existing `SponsorBeneift` configuration defines how to initialize the object. 

I'm good with the 2 models solutions since it's already what we're doing, but I just want to bring this other possibility to the table to exercise our thinking on this. Let me know what do you think about it :+1:

Reply:
@ewdurbin please, ignore this previous comment. I was able to reduce duplication using django abstract models as you can see in 8f8ac5b